compiler/src/yul/abi/functions.rs

```diff
@@ -0,0 +1,205 @@
+            } => encode_static_array(param, *inner, size),
+            AbiType::Array {
+                inner,
+                size: AbiArraySize::Dynamic,
+            } => {
+                let dyn_offset =
+                    expression! { add([static_size.clone()], [operations::sum(dyn_sizes.clone())]) };
+                let inner = *inner;
+                dyn_sizes.push(dyn_array_data_size(param.clone(), inner.clone()));
+                dyn_array_params.push((param, inner));
+                encode_uint(dyn_offset)
+            }
+            AbiType::Uint { .. } => encode_uint(param),
+        })
+        .collect::<Vec<_>>();
+
+    let dyn_encode_stmts = dyn_array_params
+        .into_iter()
+        .map(|(param, inner)| encode_dyn_array(param, inner))
+        .collect::<Vec<_>>();
+
+    return function_definition! {
+        function [func_name]([params...]) -> ptr {
+            (ptr := avail())
+            [static_encode_stmts...]
+            [dyn_encode_stmts...]
+        }
+    };
+}
+
+fn dyn_array_data_size(val: yul::Expression, inner: AbiType) -> yul::Expression {
+    match inner {
+        AbiType::Array { .. } => unimplemented!(),
+        AbiType::Uint {
+            size: AbiUintSize { padded_size, .. },
+        } => {
+            let inner_padded_size = literal_expression! { (padded_size) };
+            let array_size = expression! { mload([val]) };
+            let elements_size = expression! { mul([array_size], [inner_padded_size]) };
+            expression! { add(32, (ceil32([elements_size]))) }
+        }
+    }
+}
+
+fn encode_uint(val: yul::Expression) -> yul::Statement {
+    statement! { pop((alloc_mstoren([val], 32))) }
+}
+
+fn encode_dyn_array(val: yul::Expression, inner: AbiType) -> yul::Statement {
+    let array_size = expression! { mload([val.clone()]) };
+    let array_start = expression! { add([val], 32) };
+    block_statement! {
+        [encode_uint(array_size.clone())]
+        [encode_array(array_start, inner, array_size)]
+    }
+}
+
+fn encode_static_array(val: yul::Expression, inner: AbiType, size: usize) -> yul::Statement {
+    let array_size = literal_expression! { (size) };
+    encode_array(val, inner, array_size)
+}
+
+fn encode_array(
+    val: yul::Expression,
+    inner: AbiType,
+    array_size: yul::Expression,
+) -> yul::Statement {
+    let (inner_data_size, inner_padded_size) = match inner {
+        AbiType::Uint {
+            size:
+                AbiUintSize {
+                    data_size,
+                    padded_size,
+                },
+        } => (
+            literal_expression! { (data_size) },
+            literal_expression! { (padded_size) },
+        ),
+        AbiType::Array { .. } => unimplemented!(),
+    };
+
+    block_statement! {
+        (let array_size := [array_size])
+        (let array_data_size := mul(array_size, [inner_padded_size.clone()]))
+        (for {(let i := 0)} (lt(i, array_size)) {(i := add(i, 1))}
```

Comment:
if the inner data_size and padded_size are the same, we can just copy the whole thing into our encoding, which is much more efficient. implement that in this PR or create a separate issue for it